### PR TITLE
expression: implement vectorized evaluation for `builtinLocate2ArgsSig`

### PR DESCRIPTION
**Pull request by b41sh**
### What problem does this PR solve? 
Implement vectorized evaluation for `builtinLocate2ArgsSig`.
[12106](https://github.com/pingcap/tidb/issues/12106)

### What is changed and how it works?
```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinStringFunc/builtinLocate2ArgsSig-VecBuiltinFunc-4     3457  355022 ns/op  16720 B/op  1615 allocs/op
BenchmarkVectorizedBuiltinStringFunc/builtinLocate2ArgsSig-NonVecBuiltinFunc-4  3063  401838 ns/op  16720 B/op  1615 allocs/op
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_string_vec.go [hunk 1]
```diff
@@ -664,11 +664,52 @@ func (b *builtinLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalString(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+	result.ResizeInt64(n, false)
+	i64s := result.Int64s()
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) || buf1.IsNull(i) {
```
Comment (qw4990) [inline expression/builtin_string_vec.go]: We can use `result.MergeNulls(buf1, buf2)` here.

File: expression/builtin_string_vec_test.go [hunk 2]
```diff
@@ -29,8 +29,14 @@ var vecBuiltinStringCases = map[string][]vecExprBenchCase{
 	ast.Convert:        {},
 	ast.Substring:      {},
 	ast.SubstringIndex: {},
-	ast.Locate:         {},
-	ast.Hex:            {},
+	ast.Locate: {
+		{
+			retEvalType:   types.ETInt,
+			childrenTypes: []types.EvalType{types.ETString, types.ETString},
+			geners:        []dataGenerator{&randLenStrGener{0, 10}, &randLenStrGener{0, 20}},
```
Comment (qw4990) [inline expression/builtin_string_vec_test.go]: How about adding one more case where the length of `substrs` is between 1 to 2 to increase the probability that the `substr` is included in `str`.

Comment (b41sh) [reply expression/builtin_string_vec_test.go]: ok

File: expression/builtin_string_vec.go [hunk 3]
```diff
@@ -732,11 +732,52 @@ func (b *builtinLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+	result.MergeNulls(buf, buf1)
+	i64s := result.Int64s()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		subStr := buf.GetString(i)
+		str := buf1.GetString(i)
+
+		if int64(len([]rune(subStr))) == 0 {
```
Comment (fzhedu) [inline expression/builtin_string_vec.go]: if `subStr` is constrant,  and its length is 0, then the results are of 1.

File: expression/builtin_string_vec.go [hunk 4]
```diff
@@ -732,11 +732,52 @@ func (b *builtinLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+			i64s[i] = 1
+			continue
+		}
+		slice := string([]rune(strings.ToLower(str)))
+		ret, idx := 0, strings.Index(slice, strings.ToLower(subStr))
```
Comment (fzhedu) [inline expression/builtin_string_vec.go]: if `subStr` is constrant, please convert it to the lower case in prior.

File: expression/builtin_string_vec.go [hunk 5]
```diff
@@ -732,11 +732,54 @@ func (b *builtinLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+
+	result.ResizeInt64(n, false)
+	result.MergeNulls(buf, buf1)
+	i64s := result.Int64s()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		subStr := strings.ToLower(buf.GetString(i))
+		str := strings.ToLower(buf1.GetString(i))
```
Comment (SunRunAway) [inline expression/builtin_string_vec.go]: Could we put these two lines after the `subStrLen == 0` judgement?

Comment (b41sh) [reply expression/builtin_string_vec.go]: ok

**Pull request merged**